### DISCUSSION TITLE
Deye Speicher nur wenige BMS Werte werden angezeigt, obwohl mehr Werte in Konsole vorhanden

### DISCUSSION THREAD

**Discussion by PeterDietr** [hoylabs/OpenDTU-OnBattery#1226] [General]
Die Werte des BMS von meinem Deye ESS SE-G5.1 pro Speicher werden leider nicht alle in der Web Oberfläche angezeigt. 
In der Konsole werden einige Werte mehr übermittelt. Vermutlich weil das Deye BMS auf dem CAN-Bus nicht zu 100% Poylontech kompatibel ist.  
Kann ich das irgendwie einstellen das alle in der Konsole übermittelten Werte auf der Web Oberfläche angezeigt werden?

Aktuell werden auf der Web Oberfläche Angezeigt:
-Batterie Ladezustand in % und Volt
-Batterie Leistung in Watt und Ampere

In der Konsole wurden übermittelt:
16:37:57.310 > [Pylontech] voltage: 53.199997 current: -8.400001 temperature: 22.000000
16:37:57.377 > [Pylontech] Received CAN message: 0x0150 - 14 02 AC FF E6 01 DF 03
16:37:57.535 > [Pylontech] Received CAN message: 0x0359 - 00 00 00 00 00 00 00 00
16:37:57.651 > [Pylontech] Alarms: 0 0 0 0 0 0 0
16:37:57.755 > [Pylontech] Warnings: 0 0 0 0 0 0 0
16:37:57.858 > [Pylontech] Received CAN message: 0x035C - 08 00 00 00 00 00 00 00
16:37:57.943 > [Pylontech] chargeStatusBits: 0 0 0
16:37:58.007 > [Pylontech] Received CAN message: 0x0500 - 20 04 AA 56 37 2E 30 36
16:37:58.085 > [Pylontech] Received CAN message: 0x0351 - 40 02 70 03 E8 03 E0 01
16:37:58.163 > [Pylontech] chargeVoltage: 57.600002 chargeCurrentLimitation: 88.000000 dischargeCurrentLimitation: 100.000000
16:37:58.229 > [Pylontech] Received CAN message: 0x0400 - 01 00 60 00 00 00 0F 00
16:37:58.298 > [Pylontech] Received CAN message: 0x0355 - 31 00 63 00 00 00 00 00
16:37:58.571 > [Pylontech] soc: 49 soh: 99
16:37:58.769 > [Pylontech] Received CAN message: 0x0250 - F0 00 70 FE 58 00 64 00
16:37:58.968 > [Pylontech] Received CAN message: 0x0364 - 01 00 00 00 01 00 00 00
16:37:59.053 > [Pylontech] Received CAN message: 0x0110 - 00 00 00 00 00 00 00 31
16:37:59.133 > [Pylontech] Received CAN message: 0x0200 - 04 0D 02 0D E6 00 DC 00
16:37:59.204 > [Pylontech] Received CAN message: 0x0550 - 4B 2F 06 00 91 06 06 00
16:37:59.390 > [Pylontech] Received CAN message: 0x0363 - 20 04 20 02 00 00 00 00
16:37:59.465 > [Pylontech] Received CAN message: 0x0361 - 04 0D 02 0D E6 00 DC 00
16:37:59.533 > [Pylontech] Received CAN message: 0x0371 - 70 03 E8 03 00 00 00 00
16:37:59.682 > [Pylontech] Received CAN message: 0x035E - 44 59 30 30 31 03 E8 03
16:38:05.060 > [Pylontech] Received CAN message: 0x0364 - 01 00 00 00 01 00 00 00
16:38:05.135 > [Pylontech] Received CAN message: 0x0363 - 20 04 20 02 00 00 00 00
16:38:05.315 > [Pylontech] Received CAN message: 0x0650 - 42 34 32 30 32 30 30 31
16:38:05.381 > [Pylontech] Received CAN message: 0x0361 - 05 0D 03 0D E6 00 DC 00
16:38:05.448 > [Pylontech] Received CAN message: 0x0359 - 00 00 00 00 00 00 00 00
16:38:05.535 > [Pylontech] Alarms: 0 0 0 0 0 0 0
16:38:05.631 > [Pylontech] Warnings: 0 0 0 0 0 0 0
16:38:05.726 > [Pylontech] Received CAN message: 0x0750 - 00 00 00 00 00 00 00 00
16:38:05.827 > [Pylontech] Received CAN message: 0x0150 - 15 02 AF FF E6 01 DF 03
16:38:05.926 > [Pylontech] Received CAN message: 0x0356 - D2 14 B0 FF DC 00 00 00
16:38:06.016 > [Pylontech] voltage: 53.299999 current: -8.000000 temperature: 22.000000
16:38:06.087 > [Pylontech] Received CAN message: 0x0600 - 30 30 35 30 31 30 30 30
16:38:06.166 > [Pylontech] Received CAN message: 0x0371 - 70 03 E8 03 00 00 00 00
16:38:06.239 > [Pylontech] Received CAN message: 0x0355 - 31 00 63 00 00 00 00 00
16:38:06.339 > [Pylontech] soc: 49 soh: 99
16:38:06.405 > [Pylontech] Received CAN message: 0x035E - 44 59 30 30 31 03 E8 03
16:38:11.775 > [Pylontech] Received CAN message: 0x0110 - 00 00 00 00 00 00 00 31
16:38:11.899 > [Pylontech] Received CAN message: 0x0600 - 30 30 35 30 31 30 30 30
16:38:12.008 > [Pylontech] Received CAN message: 0x0371 - 70 03 E8 03 00 00 00 00
16:38:12.103 > [Pylontech] Received CAN message: 0x0250 - F0 00 70 FE 58 00 64 00
16:38:12.170 > [Pylontech] Received CAN message: 0x0400 - 01 00 60 00 00 00 0F 00
16:38:12.237 > [Pylontech] Received CAN message: 0x0200 - 06 0D 03 0D E6 00 DC 00
16:38:12.478 > [Pylontech] Received CAN message: 0x0500 - 20 04 AA 56 37 2E 30 36
16:38:12.582 > [Pylontech] Received CAN message: 0x0359 - 00 00 00 00 00 00 00 00
16:38:12.914 > [Pylontech] Alarms: 0 0 0 0 0 0 0
16:38:13.046 > [Pylontech] Warnings: 0 0 0 0 0 0 0
16:38:13.127 > [Pylontech] Received CAN message: 0x0150 - 15 02 A9 FF E6 01 DF 03
16:38:13.207 > [Pylontech] Received CAN message: 0x0351 - 40 02 70 03 E8 03 E0 01
16:38:13.273 > [Pylontech] chargeVoltage: 57.600002 chargeCurrentLimitation: 88.000000 dischargeCurrentLimitation: 100.000000
16:38:13.341 > [Pylontech] Received CAN message: 0x035E - 44 59 30 30 31 03 E8 03
16:38:18.657 > [Pylontech] Received CAN message: 0x0363 - 20 04 20 02 00 00 00 00
16:38:18.742 > [Pylontech] Received CAN message: 0x0351 - 40 02 70 03 E8 03 E0 01
16:38:19.031 > [Pylontech] chargeVoltage: 57.600002 chargeCurrentLimitation: 88.000000 dischargeCurrentLimitation: 100.000000
16:38:19.106 > [Pylontech] Received CAN message: 0x0364 - 01 00 00 00 01 00 00 00
16:38:19.175 > [Pylontech] Received CAN message: 0x0355 - 31 00 63 00 00 00 00 00
16:38:19.259 > [Pylontech] soc: 49 soh: 99
16:38:19.641 > [Pylontech] Received CAN message: 0x0110 - 00 00 00 00 00 00 00 31
16:38:19.743 > [Pylontech] Received CAN message: 0x035E - 44 59 30 30 31 03 E8 03

Gruß
Peter

**Comment by schlimmchen** [comment]
Also ich sehe nicht mehr Werte in diesem Log, als jene, die du aufgelistet hast und die in der Web UI auch erscheinen. Wir sehen mehr CAN Nachrichten, das ist korrekt, aber du siehst auch, dass diese nicht dekodiert werden können, weil sie offensichtlich unbekannt sind. Daher müssten wir wissen, was das alles für nachricthentypen sind und wie sie dekodiert werden, und dann muss das implementiert werden.

Wenn du uns ein Dokument lieferst, das beschreibt, was diese Register/Adressen bedeuten und wie die Daten dazu interpretiert werden müssen, dann können wir das gerne implementieren. Eine weitere Erwartung an dich wäre, dass du dann testen hilfst und aktiv Rückmeldung gibst, denn wir haben keine Deye Batterie.

@AndreasBoehm @ranma Ich hab das Gefühl, dass wir hier schon wieder viel Code duplizieren werden, stimmt's? Es müsste doch möglich sein, eine sehr allgemeine Implementierung wie "CanBattery" zu schaffen, die je nach Provider lediglich unterscheidet, welches Mapping zu verwenden ist, um die Daten zu interpretieren. Das Mapping wäre dann Register -> { Datenbeschreibung, Name, Zielvariable, etc. }. Macht das Sinn? Ich finds nämlich langsam gruselig, dass wir den ganzen Code für die Weiterleitung der Daten zum live view und zum broker und zum HASS duplizieren.

**Reply by AndreasBoehm** [reply]
Hab auch schon drüber nachgedacht wie wir eine `CanBattery` schaffen können. Sollte eigentlich nicht all zu schwierig sein. Im Bezug auf die stats können wir einfach versuchen mehr in die generischen Stats zu packen und dann noch ein paar mehr `isValidXXX()` checks hinzufügen.

**Comment by AndreasBoehm** [comment]
@PeterDietr bietet deine Batterie eine Konfiguration welches Protokoll sie nutzen soll bzw mit welchem Inverter/System sie kommunizieren soll?

**Comment by AndreasBoehm** [comment]
Hab hier infos über das Protokoll gefunden, gibt aer nicht wirklich mehr her als was wir bereits parsen: https://www.studocu.com/row/document/%E5%8D%8E%E5%8D%97%E7%90%86%E5%B7%A5%E5%A4%A7%E5%AD%A6/physical-chemistry/can-bus-protocol-%E5%BE%B7%E4%B8%9Adeye-low-voltage/46210190

**Comment by PeterDietr** [comment]
Hallo Schlimmchen, Hallo Andreas, vielen lieben Dank für die ausführliche Antwort und das Interesse daran die Deye Batterie möglicherweise zu implementieren. Gerne kann ich das Ganze dann mit den mir zur Verfügung stehenden Mitteln testen. Zu den Registern/ Adressen kann ich auch nicht mehr sagen, ich werde mal den deutschen Deye Support anschreiben und fragen ob die mir eine Dokumentation der CAN / RS485 Daten zur Verfügung stellen können.
Die Batterie bietet keine Hardware Konfiguration welches Protokoll sie nutzen soll bzw. mit welchem Inverter/System sie kommunizieren soll. Also keine Dip Schalter wie z.B. bei Pylontech.
Im Konsolen Log sehe ich die Temperatur, die aktuelle Spannung, den aktuellen Strom, den soc und soh Wert, max. Ladespannung, max. Entladeschtrom, max. Ladestrom.
In der Live Ansicht wird das Battery Statusfenster in dem die oben genannten Werte angezeigt werden könnten bei bestehender CAN Verbindung nicht angezeigt. Ohne CAN Verbindung wird in der Live Ansicht das Battery Status Fenster angezeigt, natürlich dann ohne eingetragene Batteriewerte.
Mit Can Verbindung sieht bei mir die Live Ansicht so aus. 
![Live Battery](https://github.com/user-attachments/assets/f8aa14f2-bada-48c3-a59f-4cc692f59fc4)

Von mir genutzte Hardware: ESP32-S3 USB 16MB Dev-Kit, MCP2515/Huawei PSU, und SN65HvD230VP230/Deye ESS SE-G5.1 pro
Firmware auf dem ESP: 18e2029

**Reply by AndreasBoehm** [reply]
Kannst du versuchen das Problem mit der Zeitsynchronisierung zu lösen? Ich könnte mir vorstellen das es damit zusammen hängt, bin mir aber nicht ganz sicher.

Kannst du in deinem Browser prüfen ob Daten über den websocket für livedata kommen?

Und könntest du noch einen Screenshot ohne CAN Verbindung machen?

**Reply by ranma** [reply]
> In der Live Ansicht wird das Battery Statusfenster in dem die oben genannten Werte angezeigt werden könnten bei bestehender CAN Verbindung nicht angezeigt

Ist die Firmware mit platformio selbst kompiliert? Wenn ja must du die webapp noch neu bauen, die vorkompilierte unter webapp_dist ist nicht aktuell: https://opendtu-onbattery.net/firmware/compile_webapp/

**Comment by PeterDietr** [comment]
Hier der Screenshot ohne CAN Verbindung. 
![grafik](https://github.com/user-attachments/assets/2f5e9184-c769-444c-ab59-d759913fe44c)

Das Uhrzeitproblem ist behoben, ich hatte eine feste IP Adresse in der DTU eingestellt, da funktioniert die Zeitsynchronisierung bei mir nicht. Mit aktiven DHCP funktioniert die Zeitsynchronisierung.
Das Statusfenster von der Batterie wird leider trotzdem nicht angezeigt. 

Und jetzt bin ich leider schon etwas verlohren....
Wie kann ich prüfen ob in meinem Browser Daten über den websocket für livedata kommen?
die Daten vom Shelly 3em werden alle 2 Sekunden in der Live Ansicht aktualisiert,....

@ ranma, nein ich habe nicht`s selbst kompiliert.

**Reply by AndreasBoehm** [reply]
> Um in Webbrowsern Websocket-Nachrichten nachvollziehen zu können, muss man in den Entwicklerwerkzeugen des Webbrowsers die Websocket-Ansicht (z. B. in Chrome: "WS") wählen (s. Abb. 4).

https://heise.cloudimg.io/v7/_www-heise-de_/imgs/18/3/0/6/5/2/9/0/abb4-3b04e5a5b9776e1e.png?org_if_sml=1&q=70&width=1179

**Comment by AndreasBoehm** [comment]
Kannst du dein PIN Mapping zur verfügung stelllen und nochmal Beschreiben was du alles an deinem ESP angeschlossen hast und vorallem auch wie?

**Comment by PeterDietr** [comment]
Hallo Andreas,
angeschlossen an den ESP32-S3 ist ein MC2515 Board für die CAN Kommunikation mit dem Huawei Netzteil, SN65HvD230VP230 Board für CAN Kommunikation mit dem Deye Speicher und einem SH1106 Display.  Die Werte des Netztbezuges kommen von einem Shelly 3EM. Eine Kommunikation zu Wechselrichtern besteht nicht. Ich nutze aktuell OoenDTUon Battery ausschließlich zum Überschußladen mittels dem Huawei Netzteil.

Pin Mapping:
```json
[
    {
        "name": " Esp32-S3 mit R4850G2 ; SH1106; Deye Battery ",
        "huawei": {
            "clk": 10,
            "cs": 11,
            "irq": 14,
            "miso": 13,
            "mosi": 12,
            "power": 36
        },     
        "display": {
            "clk": 4,
            "data": 5,
            "type": 3
        },
        "battery": {
            "rx": 1,
            "tx": 2
        }
    }   
]
```

**Comment by PeterDietr** [comment]
Sind das die Websocket Nachrichten die ihr benötigt?

![grafik](https://github.com/user-attachments/assets/ea93cec7-ed84-445d-bcc5-cd11724572f3)

**Comment by PeterDietr** [comment]
Eben erst gesehen, dort gibt es eine Fehlermeldung

![grafik](https://github.com/user-attachments/assets/d33798e4-cb5c-4996-8bfb-3c6f3bdb8d69)

**Reply by AndreasBoehm** [reply]
Schlechte WLAN Verbindung des ESP oder deines Rechners?

**Comment by PeterDietr** [comment]
Hallo Andreas, ich denke nein.
Das gleiche Bild habe ich wenn ich mich mit dem Handy über den AP mit dem ESP verbinde und dabei nur 1Meter von dem ESP weg bin. Die Live Daten wie z.B. Netzbezug werden auch laufend weiter aktualisiert. Die Fehlermeldung kommt nur wenn ich den Speicher am CAN angeschlossen habe.

**Comment by PeterDietr** [comment]
![grafik](https://github.com/user-attachments/assets/932b60df-fcda-453e-b0d4-6e5cdb072381)

**Reply by schlimmchen** [reply]
Interessant. Versuch bitte mal herauszufinden, was exakt da über den websocket an Daten kommt. Siehe Andreas Link oben.

**Reply by ranma** [reply]
```
16:38:13.341 > [Pylontech] Received CAN message: 0x035E - 44 59 30 30 31 03 E8 03
16:38:18.657 > [Pylontech] Received CAN message: 0x0363 - 20 04 20 02 00 00 00 00
```

Da fehlt auch irgendwie die Logmeldung `[Pylontech] Manufacturer:[...]`.

In der Tat enthält der String dann Kontrollzeichen (`DY001\x03\xe8\x03`).

Vieleicht sollte PylontechCanReceiver.cpp da nur die ersten fünf Zeichen nehmen, statt dem data_length_code zu vertrauen...

**Comment by PeterDietr** [comment]
Leider hilft mir der Link von Andreas nicht, chrome hängt sich bei mir auf dem Rechner immer auf...... als Browser nutze ich Firefox und hätte noch Edge und Opera zur Auswahl. Bei allen 3 Browsern habe ich bisher nicht herausfinden können wie ich da an die websocket Daten komme. :-(

**Comment by ranma** [comment]
Mit einem fix funktioniert das hier mit einem replay deiner CAN-Nachrichten:
![image](https://github.com/user-attachments/assets/988a61a2-5689-4c37-98cf-74b3850ad2e2)

Pull request: https://github.com/helgeerbe/OpenDTU-OnBattery/pull/1235

**Reply by PeterDietr** [reply]
@ ranma, wow das sieht doch richtig gut aus :-)  wie bekomme ich zum testen deinen Fix auf meinen ESP?

**Reply by ranma** [reply]
Schau mal hier unter "Artifacts": https://github.com/helgeerbe/OpenDTU-OnBattery/actions/runs/10741708550?pr=1235

**Reply by PeterDietr** [reply]
Per OTA in den ESP übertragen und... yes es funktioniert :-)
Ganz gang vielen lieben Dank an alle beteiligten die sich die Zeit genommen und Mühe gegeben haben das "Problemchen" zu lösen. 
Ich habe von meiner besseren Hälfte für heute leider OpenDTU Verbot bekommen......
Anfang nächster Woche werde ich mit den Neuerungen Spielen und Berichten.
Danke den Beteiligten und ein schönes WE ;-)

![grafik](https://github.com/user-attachments/assets/920f2958-de65-458e-be76-b0d56522e195)

**Reply by AndreasBoehm** [reply]
@PeterDietr weißt du ob dein Akku zum Zeitpunkt des Screenshots gerade geladen oder entladen wurde?
Laut CAN Protocol ist bei Deye die Angabe des Stroms anders als bei anderen Herstellern.
**Positiv** -> Entladen
**Negativ** -> Laden

**Comment by schlimmchen** [comment]
@PeterDietr @AndreasBoehm @ranma Die WebApp wird mit https://github.com/helgeerbe/OpenDTU-OnBattery/commit/cdd6f4cf1b9c08dfd871630ab201e10e21335030 behaupten, dass der Pylontech Battery Provider auch DEYE versteht. Ist das korrekt so oder habe ich etwas missverstanden?

**Reply by AndreasBoehm** [reply]
Ich denke soweit würde ich micht gehen. Es funktioniert teilweise. Im Screenshot ist zu sehen das “Laden ermöglicht” und “Entladen ermöglicht” auf “Nein” steht obwohl ja alles funktioniert.

Das verhält sich ähnlich wie eine Pytes Batterie die man als Pylontech in der DTU konfiguriert, es werden werte geparst aber lange nicht alle und wie bei “Laden ermöglicht” werden nur die defaults angezeigt.

Vermutlich werden auch Warnings und Alarms nicht funktionieren.

**Reply by schlimmchen** [reply]
Hm. Warum tauchen diese Zeilen dann auf im LiveView? In dieser Tabelle sollten nur Zeilen stehen, die ausdrücklich übermittelt werden, also wurden die Daten auch ausgesucht zum übermitteln. Wenn die gar nie empfangen wurden... warum ist dann da eine Zeile in der Tabelle?

Verstehe... Das sind einfach `bool`s in `PylontechBatteryStats` und die werden übermittelt, egal ob sie jemals gesetzt wurden, oder nicht.

Also das ließe sich leicht heilen, indem das `std::optional<bool>` werden.

Aber dann fehlt noch, dass da steht "No Issues" in der web app, obwohl sehr wohl Issues gemeldet sein können, die wir nicht verstehen? Wie wäre es da mit einer Meldung "Warnung: Nicht unterstützt" bis wir die erste bekannte Nachricht der Batterie dekodieren und dann sagen können, ob Alarme und Warnungen gesetzt sind?

Fang ich an zu arg zu basteln, oder wäre das aus deiner/euer Sicht akzeptabel, um Deye Batterien ausdrücklich zu unterstützen?

**Reply by AndreasBoehm** [reply]
Optionals einzusetzen halte ich für eine sehr gute Idee. Damit können wir in Zukunft in `BatteryStats` auch mehr Werte generalisieren und müssen nicht so viel code duplizieren.

Das Thema mit den Issues muss ich mir nochmal anschauen, laut den Logs wird das für Deye ordentlich geparsed aber eventuell nutzt Deye trotzdem ein anderes Format. Grundsätzlich finde ich die Idee aber super hier auch eine art optional Handling zu haben.

**Reply by schlimmchen** [reply]
Ich reverte meinen Commit, und wir tracken das Thema in #1244. Bis dahin ist die Unterstützung für DEYE Batterien aus den von Andreas genannten Gründen nicht ausdrücklich gegeben.

**Reply by PeterDietr** [reply]
Kurze Rückmeldung von mir. Bisher läuft die Kommunikation mit dem Deye BMS ohne erkennbare Probleme. 
Die Anfrage an Deye wegen dem CAN-Protokoll habe ich verschickt, leider noch keine Antwort erhalten. Richtig ist das "Laden ermöglicht" ; "Entladen ermöglicht" und "Sofortiges Laden angefordert" immer auf "Nein" stehen, der Akku aber ganz normal geladen und entladen werden kann.  Ob Fehler oder Warnungen richtig an die DTU übermittelt werden kann ich schlecht testen, da der "Normalbetrieb" ja keine Fehler oder Warnungen erzeugt.

**Comment by PeterDietr** [comment]
Ich hab evon Deye das Can Bus Protokoll erhalten :-)

Da ich nicht weis ob Deye damit einverstanden ist es hier öffentlich einzustellen.....
Wie kann ich das File Euch zukommen lassen?

**Reply by AndreasBoehm** [reply]
Gerne per email an andreas@boehm.cx

**Comment by Adminius** [comment]
Ich habe Deye Batterien Protokoll, zwar anders Modell, scheint aber gleich zu sein, fast vollständig dekodiert, falls es jemanden interessiert:
https://github.com/Adminius/esphome-yaml-collection/blob/main/deye_rw-m6.1.yaml
